src/main/java/org/folio/service/mapping/writer/impl/AbstractRecordWriter.java

```diff
@@ -0,0 +1,105 @@
+package org.folio.service.mapping.writer.impl;
+
+import org.folio.service.mapping.processor.rule.DataSource;
+import org.folio.service.mapping.reader.values.CompositeValue;
+import org.folio.service.mapping.reader.values.ListValue;
+import org.folio.service.mapping.reader.values.SimpleValue;
+import org.folio.service.mapping.reader.values.StringValue;
+import org.folio.service.mapping.writer.RecordWriter;
+import org.folio.service.mapping.writer.fields.RecordControlField;
+import org.folio.service.mapping.writer.fields.RecordDataField;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.folio.service.mapping.reader.values.SimpleValue.SubType.LIST_OF_STRING;
+import static org.folio.service.mapping.reader.values.SimpleValue.SubType.STRING;
+
+public abstract class AbstractRecordWriter implements RecordWriter {
+  private static final String INDICATOR_1 = "1";
+  private static final String INDICATOR_2 = "2";
+
+  @Override
+  public void write(SimpleValue simpleValue) {
+    DataSource dataSource = simpleValue.getDataSource();
+    String tag = dataSource.getTag();
+    if (STRING.equals(simpleValue.getSubType())) {
+      StringValue stringValue = (StringValue) simpleValue;
+      if (dataSource.isSubFieldSource() || dataSource.isIndicatorSource()) {
+        RecordDataField recordDataField = buildDataFieldForStringValues(tag, Collections.singletonList(stringValue));
+        writeDataField(recordDataField);
+      } else {
+        RecordControlField recordControlField = new RecordControlField(tag, stringValue.getValue());
+        writeControlField(recordControlField);
+      }
+    } else if (LIST_OF_STRING.equals(simpleValue.getSubType())) {
+      ListValue listValue = (ListValue) simpleValue;
+      if (dataSource.isSubFieldSource() || dataSource.isIndicatorSource()) {
+        RecordDataField recordDataField = buildDataFieldForListOfStrings(listValue);
+        writeDataField(recordDataField);
+      } else {
+        for (String value : listValue.getValue()) {
+          RecordControlField recordControlField = new RecordControlField(tag, value);
+          writeControlField(recordControlField);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void write(CompositeValue compositeValue) {
+    String tag = compositeValue.getValue().get(0).get(0).getDataSource().getTag();
+    for (List<StringValue> entry : compositeValue.getValue()) {
```

Comment:
I think we should call the  write(SimpleValue simpleValue)  here for each value, because each type can then have string or array types isn't it?

Reply:
Yes, we can, but lets this logic separated. Some of these places may be changed after rework in future